ngraph/core/reference/include/ngraph/runtime/reference/gather.hpp

```diff
@@ -16,158 +16,116 @@
-                std::iota(indices_outer_axis_order.begin(), indices_outer_axis_order.end(), 0);
-                const CoordinateTransform indices_outer_transform(indices_shape,
-                                                                  indices_outer_start_corner,
-                                                                  indices_outer_end_corner,
-                                                                  indices_outer_strides,
-                                                                  indices_outer_axis_order);
-
-                // Create an inner CoordinateTransfrom for "out"
-                const size_t out_inner_ndim = out_ndim - axis;
-                const Shape out_inner_shape(out_shape.begin() + axis, out_shape.end());
-                const Coordinate out_inner_start_corner(out_inner_ndim, 0);
-                Coordinate out_inner_end_corner(out_inner_shape);
-                if (indices_ndim > 0)
-                {
-                    out_inner_end_corner[indices_ndim - 1] = 1;
-                }
-                for (size_t i = indices_ndim; i < out_inner_ndim; i++)
-                {
-                    out_inner_end_corner[i] = 1;
-                }
-                const Strides out_inner_strides(out_inner_ndim, 1);
-                AxisVector out_inner_axis_order(out_inner_ndim);
-                std::iota(out_inner_axis_order.begin(), out_inner_axis_order.end(), 0);
-                const CoordinateTransform out_inner_transform(out_inner_shape,
-                                                              out_inner_start_corner,
-                                                              out_inner_end_corner,
-                                                              out_inner_strides,
-                                                              out_inner_axis_order);
-
-                auto out_outer_coord_iter = out_outer_transform.begin();
-                for (const Coordinate& params_outer_coord : params_outer_transform)
+                const auto params_axes_part = span(params_shape).subspan(0, axis);
+
+                NGRAPH_CHECK(params_shape.size() >= axis, "Not enough axes in param_shape.");
+
+                const auto reminder_part_shape = span(params_shape).subspan(axis + 1);
+
+                const auto found_out_shape =
+                    join(params_axes_part, span(indices_shape), reminder_part_shape);
+
+                NGRAPH_CHECK(found_out_shape == out_shape,
+                             "Output shape mismatch with calculations");
+
+                const auto batch_shape = span(params_shape).subspan(axis);
+
+                const auto batch_size = shape_size(batch_shape);
+
+                const auto copy_size = shape_size(reminder_part_shape);
+
+                const size_t copy_round_in_batch =
+                    indices_shape.empty() ? 1 : shape_size(indices_shape) / indices_shape.back();
```

Comment:
I'll add check here.
What's the general approach for shape with `0` on axes? How we should proceed this because `shape_size` will return 0?

Reply:
In general we should be ready to get a `0` in dimension. At the current moment I think we can throw an exception or return the error... 
But as I know `0` is a valid value in the TensorFlow models, and I think in the future we should support to work with such dimensions.
@lazarevevgeny Could you advise or add something?

Reply:
Yes, it is possible that the dimension will be equal to 0 (meaning empty tensor). In this case the corresponding output dimension should be also 0. It will be great if we can handle this case in this PR, but I don't want to insist on this since the 0 dimension tensors are not yet supported and we don't have models with such a case now.

Reply:
For clarifying. Is possible to have tensor with `rank > 1` and one of `axes[i] == 0` and is this mean 'empty tensor'? 
How tensor with `rank == 0` is distinguish from tensor with `rank > 0` and some `axiex == 0`?

Reply:
Yes, it is possible. Consider the model where you calculate bounding boxes for objects and then filter them by some probability threshold. There could be a situation when all the boxes have probability lower than threshold so the dimension size will be 0. Situation when `rank == 0` means the the number of dimensions is 0, so the tensor is scalar and its shape is `[]`. The empty tensor is a tensor of arbitrary rank ND but one of the dimension is equal to 0, for example, the possible shape of empty tensor could be `[100, 0, 123, 454]`.

Reply:
1) Is empty tensor means 'no data' (for now `shape_size` for `shape` of empty tensor return `0`)?
2) In gather op (and probably many others ops) if output tensor is empty tensor operator has noting to do?

Reply:
1. Yes
2. Yes